src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -12,100 +12,473 @@
+  // a new state object.
+  let state = prevState;
+  let dontMutatePrevState = true;
+  let isEmpty = true;
+  let callbackList = null;
+  let update = queue.first;
+  while (update && comparePriority(update.priorityLevel, priorityLevel) <= 0) {
+    // Remove each update from the queue right before it is processed. That way
+    // if setState is called from inside an updater function, the new update
+    // will be inserted in the correct position.
+    queue.first = update.next;
+    if (!queue.first) {
+      queue.last = null;
+    }
+
+    let partialState;
+    if (update.isReplace) {
+      // A replace should drop all previous updates in the queue, so
+      // use the original `prevState`, not the accumulated `state`
+      state = getStateFromUpdate(update, instance, prevState, props);
+      dontMutatePrevState = true;
+      isEmpty = false;
+    } else {
+      partialState = getStateFromUpdate(update, instance, state, props);
+      if (partialState) {
+        if (dontMutatePrevState) {
+          state = Object.assign({}, state, partialState);
+        } else {
+          state = Object.assign(state, partialState);
+        }
+        dontMutatePrevState = false;
+        isEmpty = false;
+      }
+    }
+    if (update.isForced) {
+      workInProgress.effectTag |= ForceUpdate;
+    }
+    if (update.callback) {
+      if (callbackList && callbackList.last) {
+        callbackList.last.next = update;
+        callbackList.last = update;
       } else {
-        callback();
+        callbackList = {
+          first: update,
+          last: update,
+        };
       }
+      workInProgress.effectTag |= CallbackEffect;
     }
-    node = node.next;
+    update = update.next;
   }
-};
 
-function getStateFromNode(node, instance, state, props) {
-  if (typeof node.partialState === 'function') {
-    const updateFn = node.partialState;
-    return updateFn.call(instance, state, props);
-  } else {
-    return node.partialState;
+  if (isEmpty) {
+    // None of the updates contained state. Use the original state object.
+    state = prevState;
   }
-}
 
-exports.mergeUpdateQueue = function(queue : UpdateQueue, instance : any, prevState : any, props : any) : any {
-  let node : ?UpdateQueueNode = queue;
-  if (queue.isReplace) {
-    // replaceState is always first in the queue.
-    prevState = getStateFromNode(queue, instance, prevState, props);
-    node = queue.next;
-    if (!node) {
-      // If there is no more work, we replace the raw object instead of cloning.
-      return prevState;
-    }
+  if (!queue.first) {
+    // Queue is now empty
+    workInProgress.updateQueue = null;
   }
-  let state = Object.assign({}, prevState);
-  while (node) {
-    let partialState = getStateFromNode(node, instance, state, props);
-    Object.assign(state, partialState);
-    node = node.next;
+
+  workInProgress.callbackList = callbackList;
+  workInProgress.memoizedState = state;
```

Comment:
I have concerned because we don't set memoizedProps here. If we abort and then resume we might end up relying on the wrong memoizedState but I don't see any bugs yet and we probably want to move to a model where memoizedProps is also set in begin.